src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -1500,10 +1523,27 @@ internal static int IndexOfChar(ref char searchSpace, char value, int length)
                 {
                     length -= 1;
 
-                    if (TNegator.NegateIfNeeded(Unsafe.Add(ref searchSpace, offset) == value)) return (int)offset;
+                    if (TNegator.NegateIfNeeded(Unsafe.Add(ref searchSpace, offset) == value)) goto Found;
 
                     offset += 1;
                 }
+                return -1;
+            Found7:
+                return (int)(offset + 7);
```

Comment:
do we really need these gotos? One of the primary goals of https://github.com/dotnet/runtime/pull/73768 was to simplify the code and get rid of the gotos (I am writing guidelines for how to use `Vector12/256` and my job was to verify that users can write simple but performant code using the new APIs). What difference do the benchmarks report when using goto? If the difference is worth it, could you please log a codegen issue that would allow us to get rid of them?

Reply:
@adamsitnik Consider for example this very simple benchmark: https://gist.github.com/BrzVlad/59ab5dea7b8dbbd632877f66882c1c6e

On both mono jit and mono interp the goto version is significantly faster (about 35%). Here is the diff on mono interp https://www.diffchecker.com/hsvgFTPt. As described in the commit, the generated code for the hot loop with the goto is more compact and with less branches so it is faster.

On coreclr there is no diference in performance, which I expected. Looking at the generated code is a little bit creepy since the the codegen for both methods is absolutely identical https://www.diffchecker.com/GhauFh1f.